backend/de.metas.contracts/src/main/java/de/metas/contracts/bpartner/repository/OrgChangeRepository.java

```diff
@@ -0,0 +1,734 @@
+				.nonMembershipSubscriptions(getNonMembershipSubscriptions(bpartnerId, orgChangeDate, bPartnerComposite.getOrgId()))
+				.build();
+	}
+
+	private Collection<? extends FlatrateTerm> getNonMembershipSubscriptions(final BPartnerId bpartnerId, final LocalDate orgChangeDate, final OrgId orgId)
+	{
+		final Set<FlatrateTermId> flatrateTermIds = retrieveNonMembershipSubscriptions(bpartnerId, orgChangeDate, orgId);
+		return flatrateTermIds.stream()
+				.map(this::createFlatrateTerm)
+				.collect(ImmutableList.toImmutableList());
+	}
+
+	private List<FlatrateTerm> getMembershipSubscriptions(final BPartnerId bpartnerId, final LocalDate orgChangeDate, final OrgId orgId)
+	{
+		final Set<FlatrateTermId> membershipFlatrateTermIds = retrieveMembershipSubscriptions(bpartnerId, orgChangeDate, orgId);
+		return membershipFlatrateTermIds.stream()
+				.map(this::createFlatrateTerm)
+				.collect(ImmutableList.toImmutableList());
+	}
+
+	private FlatrateTerm createFlatrateTerm(@NonNull final FlatrateTermId flatrateTermId)
+	{
+		final I_C_Flatrate_Term term = flatrateDAO.getById(flatrateTermId);
+
+		final OrgId orgId = OrgId.ofRepoId(term.getAD_Org_ID());
+
+		final ZoneId timeZone = orgDAO.getTimeZone(orgId);
+
+		return FlatrateTerm.builder()
+				.flatrateTermId(flatrateTermId)
+				.orgId(orgId)
+				.billPartnerID(BPartnerId.ofRepoId(term.getBill_BPartner_ID()))
+				.billLocationId(BPartnerLocationId.ofRepoIdOrNull(term.getBill_BPartner_ID(), term.getBill_Location_ID()))
+				.shiToBPartnerId(BPartnerId.ofRepoIdOrNull(term.getDropShip_BPartner_ID()))
+				.shipToLocationId(BPartnerLocationId.ofRepoIdOrNull(term.getDropShip_BPartner_ID(), term.getDropShip_Location_ID()))
+				.productId(ProductId.ofRepoId(term.getM_Product_ID()))
+				.flartareConditionsId(ConditionsId.ofRepoId(term.getC_Flatrate_Conditions_ID()))
+				.isSimulation(term.isSimulation())
+				.status(FlartrateTermStatus.ofNullableCode(term.getContractStatus()))
+				.userId(UserId.ofRepoIdOrNull(term.getAD_User_InCharge_ID()))
+				.startDate(TimeUtil.asLocalDate(term.getStartDate(), timeZone))
+				.endDate(TimeUtil.asLocalDate(term.getEndDate(), timeZone))
+				.masterStartDate(TimeUtil.asLocalDate(term.getMasterStartDate(), timeZone))
+				.masterEndDate(TimeUtil.asLocalDate(term.getMasterEndDate(), timeZone))
+				.plannedQtyPerUnit(term.getPlannedQtyPerUnit())
+				.uomId(UomId.ofRepoIdOrNull(term.getC_UOM_ID()))
+				.build();
+	}
+
+	public void saveOrgChangeBPartnerComposite(@NonNull final OrgChangeBPartnerComposite orgChangeBPartnerComposite)
+	{
+		final BPartnerComposite bPartnerComposite = orgChangeBPartnerComposite.getBPartnerComposite();
+
+		bPartnerComposite.getBpartner().setOrgMappingId(orgChangeBPartnerComposite.getBPartnerOrgMappingId());
+
+		bPartnerCompositeRepo.save(bPartnerComposite);
+
+		// TODO subscriptions
+	}
+
+	@Nullable
+	public BPartnerLocation getBillToDefaultLocationOrNull(final List<BPartnerLocation> locations)
```

Comment:
i think all the following methods (from here till `unmarkShipToDefaultContacts`) don't really belong into a repo..
i would rather see them instance methods..maybe of OrgChangeBPartnerComposite?
Also, are you aware of `BPartnerComposite#extractLocation(java.util.function.Predicate<de.metas.bpartner.composite.BPartnerLocation>)`? maybe you can use it

Reply:
You are right. Using that instead.